### ISSUE TITLE
Can't get the "※"(REFERENCE MARK) character to be *wide*

### ISSUE THREAD

**Issue by BladeMight** [mintty/mintty#973]
Normally it displays as:
![1-no-abmig](https://user-images.githubusercontent.com/11931888/76125680-ffbe8380-6005-11ea-80aa-4a95bb43af75.png)
With `Charwidth=ambig-wide` it displays as:
![2-ambig-wide](https://user-images.githubusercontent.com/11931888/76125703-0c42dc00-6006-11ea-8ee6-aafa721c1f3f.png)
(this is how I'd like it to bee)
But there is a problem, some other symbols are being *widened* as well, for example Cyrillic:
![3-why-ambig-wide](https://user-images.githubusercontent.com/11931888/76125757-2a104100-6006-11ea-9daf-5896f53b8d22.png)
because of `Charwidth=ambig-wide`

I have tried:
 - `CharNarrowing=100` - no change, same cut-off char borders.
 - `CharNarrowing=50` - squashed but whole character is displayed.
 - *FontChoice* gives no changes at all:
```
Font=Consolas
FontChoice=CJK:2;Han:2;Hangul:2;PictoSymbols:2;
Font2=MS Gothic
```
- Tried to start with different locale: `LC_ALL=ja_JP.UTF-8@cjknarrow mintty -o Locale= --fn=Consolas` (with and without **\@cjknarrow**) - no changes.
- Tried other similar approaches from issues, no effect.

my locale output:

```
$ locale
LANG=C.UTF-8
LC_CTYPE="C.UTF-8"
LC_NUMERIC="C.UTF-8"
LC_TIME="C.UTF-8"
LC_COLLATE="C.UTF-8"
LC_MONETARY="C.UTF-8"
LC_MESSAGES="C.UTF-8"
LC_ALL=
```

Only the `Charwidth=ambig-wide` changes display of that character, but it changes a lot of other characters as well, so what can I do?

mintty 3.14, from msys repo.
I use Consolas font.

The characters from `\u4e00`..`...` are displayed correctly, and some from `\u2000` are displayed incorrectly:
![utf-ranges display](https://user-images.githubusercontent.com/11931888/76127747-26cb8400-600b-11ea-92d4-d115776710b9.jpg)

**Comment by mintty** [comment]
Not sure what you're up to. All behaviour you describe is correct. Character width is defined by Unicode. Charwidth=ambig-wide is *one* common variation (i.e. not per character), as also defined by Unicode.

**Comment by BladeMight** [comment]
> Not sure what you're up to

To display the `※` as wide character(like in `ambig-wide` mode), just like `一` and etc. without `ambig-wide` mode, but don't display characters like `АбвгД` as wide.

**Comment by mintty** [comment]
There is no standard mode available that would support this distinction, as character width is *normally* determined via the locale API.
However, mintty does in fact already support modes for wide display of Indic characters or "long" characters, respectively, switchable by an escape sequence. See https://github.com/mintty/mintty/wiki/CtrlSeqs#wide-characters.
We can talk about extending the set of "long" characters, but then I'd need a solid proposal by which criteria any character should be included for that mode.

**Comment by BladeMight** [comment]
> which criteria

I think the ones that going out of character borders, like `※`. And you already did that for the japanese/chinese symbols.

> determined via the locale API

Does that mean that this character would be displayed same on "normal" linux terminal with my current locale(or en_US.utf8 or ja_JP.utf8)?

> https://github.com/mintty/mintty/wiki/CtrlSeqs#wide-characters.

I absolutely don't get how to use them, `echo`in them does nothing...?

**Comment by BladeMight** [comment]
> I absolutely don't get how to use them, echo'ing them does nothing...?

I get it, but not for the `^[]77119;2^G` it says:

> can be used to change wide display modes for Indic and a range of long Unicode characters

But how can I specify those ranges here?

**Comment by BladeMight** [comment]
btw, isn't the `※` part of CJK?

**Comment by mintty** [comment]
The character is in the Unicode range General Punctuation (U+2000-U+206F).
It is marked to have "ambiguous width" in Unicode data, so it will be wide in respective width modes, along with a range of other characters as you originally observed, and narrow otherwise.

> I think the ones that going out of character borders, like ※.

You cannot generally predict whether a *glyph* (i.e. the graphic representation in a font) for a certain character is wider than something. A similar request has been discussed for "Powerline symbols" and others, but that is not a feasible approach.

> And you already did that for the japanese/chinese symbols.

No. Mintty follows Unicode width specifications, there is no extra handling for CJK symbols.

> > determined via the locale API

> Does that mean that this character would be displayed same on "normal" linux terminal with my current locale(or en_US.utf8 or ja_JP.utf8)?

Sure, unless that terminal is buggy.

> But how can I specify those ranges here?

You can make a proposal *what* characters to make wide additionally, but that would need to be based on reasonable criteria for any character, not just one particular one.

**Comment by BladeMight** [comment]
> glyph (i.e. the graphic representation in a font)

You're right, I totally forgot that fonts too have their own character sizes...

Maybe you could let users define their own characters/ranges that should/or not be double width?

Although it is in `General Punctuation` range, its width is more similar to CJK's `一`, just like `⁂`, so far I have seen problems with only these two symbols(or should i say glyphs of MS Gothic font of these symbols?).

**Comment by mintty** [comment]
You can prevent the clipping with option `CharNarrowing=50`.

**Comment by mintty** [comment]
The ASTERISM U+2042 is even clearly in the narrow category of Unicode, so I guess your wish will stay void unless you would define a new locale AND get a generic locale mechanism added to cygwin.
Marking invalid.

**Comment by BladeMight** [comment]
@mintty well if thats the case... btw how are these symbols displayed in your configuration?

**Comment by mintty** [comment]
Font DejaVu Sans Mono 12pt, locale en_GB.UTF-8:
![image](https://user-images.githubusercontent.com/12740416/76409725-0d219800-638f-11ea-94c1-caa63d0d39cb.png)

**Comment by BladeMight** [comment]
That looks OK, with for me with same font only on 12pt the results are OK:
![2020 03 11 - 13'13'26_rect](https://user-images.githubusercontent.com/11931888/76411503-a2765980-639a-11ea-9829-daa93bfafa84.jpg)
with `CharNarrowing=70` `Consolas 11pt` it looks almost OK:
![2020 03 11 - 13'18'55_rect](https://user-images.githubusercontent.com/11931888/76411716-0ac53b00-639b-11ea-810b-913c35e76d1d.jpg)

Looks like Consolas font characters are too tall...

**Comment by mintty** [comment]
Released 3.1.5 with explicit character width attribute support.

**Comment by BladeMight** [comment]
Good, now we can make characters wide whenever we need, like that: `printf "\033[1 Z※\033[0 Z"`. Thanks.
